Batch up all `touch` calls made in a transaction.

BMorearty opened the pull request:
Greatly reduces the number of database round-trips made when `touch` is called on a lot of records in a transaction. It does so by delaying the `touch` calls until just before the transaction is committed. At that point, it consolidates the touches into the smallest possible number of SQL UPDATEs by (a) only touching each record once, even if it was touched repeatedly in the block, and (2) touching multiple records of a single table simultaneously, using an IN clause.

`after_touch` calls are also deferred until right after the database updates occur, just before the commit.

The original code is from the [activerecord-delay_touching](https://github.com/godaddy/activerecord-delay_touching/) gem, which has been in production on multiple sites for a bit over six months.

Some benchmarks: 
1. I wrote a sample app where `Person has_many :pets` and `Pet belongs_to :person, touch: true`. So each time a pet is updated, the person who owns it is touched. I made a person who has 25 pets. Then I ran this code in Rails 5 with and without this change:
   
   ``` ruby
   Benchmark.ips do |x| 
     ids = person.pet_ids
     x.report do
       h = ids.map { |id| { id:id, name:rand } }
       # Update all 25 pets. Touches person 25x without the change, 1x with it.
       person.update pets_attributes: h 
     end
   end
   ```
   
   The result was **52 iterations per second** without this change and **91 iterations per second** with it.
2. One of the production sites where activerecord-delay_touching has been in use is a Spree site. To benchmark this change, @mtuckergd created 6 products. He then added 75 variants to each product. He benchmarked code that processes changes to all of the products' variants and updates the master stock count for each of them. The Spree models that got saved were `Variant`, `StockItem`, and `StockMovement`. Three runs were with delay touching disabled, and three were with it enabled. Results:
   
   | Without delay touching (seconds) | With delay touching (seconds) |
   | --- | --- |
   | 21.27 | 10.37 |
   | 20.26 | 10.21 |
   | 21.22 | 10.45 |
   | **Average:** |  |
   | 20.92 | 10.34 |
   
   That’s a 51% improvement to the average processing time.

Tests are included for edge cases such as rolling back a nested transaction, both with and without `:requires_new`.

Fixes rails/rails#18606.

BMorearty committed:
Batch up all `touch` calls made in a transaction.

Greatly reduces the number of database round-trips made when `touch` is
called on a lot of records.

Fixes rails/rails#18606.

dmitry commented on activerecord/test/schema/schema.rb:
```diff
@@ -499,6 +499,7 @@ def except(adapter_names_to_exclude)
     t.string :name
     t.column :updated_at, :datetime
     t.column :happy_at,   :datetime
+    t.column :sad_at  ,   :datetime
```
Space between comma.

BMorearty replied:
Whoops, I'll fix this.

dmitry commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+
+      def add_record(record, columns)
+        # Include the standard updated_at column and any additional specified columns
+        updated_at_attrs = record.send(:timestamp_attributes_for_update_in_model)
```
Access to the private method. Must use some kind of the getter.

BMorearty replied:
Good catch. I can change `timestamp_attributes_for_update_in_model` to a public method with `:nodoc:`. That way it will be callable from outside the record, but not documented as a part of the public ActiveRecord API.

dmitry commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,119 @@
+      states.last
+    end
+
+    class << self
```
Why not using `class << self` everywhere, as in `NoTouching` module?

BMorearty replied:
That was an oversight on my part. I wrote the original code while on a team where the preferred style was `self.` and I only used `class << self` in places where it's required, such as the `delegate` declaration.

I'll fix the style to use the Rails convention.

dmitry commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,119 @@
+    # Touch the specified records--non-empty set of instances of the same class.
+    def self.touch_records(klass, attrs, records)
+      if attrs.present?
+        current_time = records.first.send(:current_time_from_proper_timezone)
```
Second time access to the same private method.

BMorearty replied:
Actually it's a different method this time, but your point is valid. :smile: I'll make it public and mark it with `:nodoc:`.

arthurnn commented:
(Please correct me if I am wrong). As far as I can tell the way you track this, is holding a reference to all records that call `.touch` inside a transaction. The problem about this implementation is, that GC wont collect them, so something like this would inflate memory a lot:
```ruby
Post.transaction do
  100_000.times do
    Post.last.touch
  end
end
```

cc @tenderlove @jeremy

BMorearty commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,119 @@
+          end
+        end
+
+        klass.unscoped.where(klass.primary_key => records.sort).update_all(changes)
```
I noticed one change I forgot to make, that I have to make before this PR is merged: increment the optimistic locking column for all records if locking is enabled. (See https://github.com/rails/rails/blob/master/activerecord/lib/active_record/persistence.rb#L478).

BMorearty commented:
@arthurnn You are correct. I was concerned about this as well.

It should be noted that [Rails already does this](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L57) within a transaction for all records that have `after_commit` or `after_rollback` hooks. Any such record that is created or updated within a transaction will be held in memory until the transaction is finished.

In the case of touching, I've have a proposed fix but I think it could be implemented in a separate PR if it becomes a problem. The proposal is to set a limit to the number of records that are held in the buffer, and flush them to the DB if we reach that number of records. If necessary, it could even be an option to `transaction` to let callers override the default, but I don't think that is necessary.

The only issue I can think of for such a mechanism is dealing with rollbacks consistently.

BMorearty committed:
Changes based on code review.

* Use `class << self`.
* Make private methods public, marked with :nodoc:.
* Fix spaces in schema.rb.

BMorearty commented:
Restoring a hidden comment that's still relevant: I noticed one change I forgot to make, that I have to make before this PR is merged: increment the optimistic locking column for all records if locking is enabled. (See https://github.com/rails/rails/blob/master/activerecord/lib/active_record/persistence.rb#L478).

egilburg commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,119 @@
+        end
+
+        state.updated klass, attrs, records
+        records.each { |record| record._run_touch_callbacks }
```
`records.each(&:_run_touch_callbacks)`

BMorearty committed:
Update based on code review.

egilburg commented:
Thanks for your work on this.

With all the `touch` logic splittered across multiple places (Persistence, NoTouching, DelayTouching, etc) I wonder whether it'd be more manageable to aggregate all that code into a central Touch module? (it can then still break up implementation into the above mixins or similar, but it'd be top-down structured and encapsulated to the outside, rather than bubble up layers of complexity.

Anyone else has an opinion on this?

dmitry commented:
I agree with @egilburg. It's a bit spread through the code.

BMorearty commented:
That's an interesting proposal, @egilburg. It's a tricky thing, trying to figure out the right way to organize the `touch` code. `Transaction` also overrides `touch`.

If there ends up being agreement that the `touch` code needs to be reorganized, I would strongly suggest it be done in a separate PR after this one is complete. Refactoring is easier to test, review, and understand if it's done separately from changing/adding features.

BMorearty committed:
Use SortedSet instead of Set for better performance.

Using benchmark-ips in Ruby 2.2, I found that:

1. Adding records to a SortedSet is just as fast as adding them to a Set.

2. Calling `sorted_set.to_a` and iterating over the results is 337 times
   faster than calling `set.sort` and iterating over those results.
   (42,180 vs. 125 iterations per second.) The memory usage is the same,
   because both methods create a temporary array.

BMorearty committed:
Treat touch :column and touch "column" as identical.

BMorearty committed:
Support optimistic locking.

Just as with a `touch` that occurs outside a transaction,
this code will update the optimistic locking column
but will not fail if the column was changed by another instance.
When updating instances in batch, it's impossible to use the
lock_version to determine whether another instance has
modified the record.

BMorearty commented:
I have fixed the issue I raised about incrementing the optimistic locking column. The fix is in the `touch_records` method.

dhh commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+
+      protected
+
```
:scissors: newline and indent under the privacy.

BMorearty replied:
Done.

dhh commented:
Lovely work on this. I'll let others help with the detailed review, but this will be a great addition to Rails. I've seen more than my fair share of crazy touch cascades.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+
+      delegate :add_record, to: :state
+
+      # Are we currently executing in a delay_touching block?
```
:scissors: this line

BMorearty replied:
Done.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+
+      # Are we currently executing in a delay_touching block?
+      def delay_touching?
+        DelayTouching.states.length > 0
```
In this scope `self` is `DelayTouching`, so you can rely on implicit self. Could also just be `!states.empty?`.

BMorearty replied:
Good idea. Done. (Actually, I changed it to `states.present?`.)

kaspth replied:
That'll do it :+1:

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+      ensure
+        merge_transactions unless $! && options[:requires_new]
+
+        # Decrement nesting even if +apply+ raised an error.
```
Can you document why we need to decrement the nesting `even if +apply+ raised an error`?

BMorearty replied:
Yes. Done.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+      # touched records with the outer transaction's touched records.
+      def merge_transactions
+        num_states = states.length
+        states[num_states - 2].merge!(states[num_states - 1]) if num_states > 1
```
Could this be `states[-2].merge!(state) if delay_touching?`?

BMorearty replied:
Sure, that's more clear. Done.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+        Thread.current[:delay_touching_states] ||= []
+      end
+
+      def state
```
I'd suggest renaming this to `current_state`. I keep thinking it's a local variable in some of the methods below when `state` is just there.

BMorearty replied:
Good idea. `current_state` is a better name.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,118 @@
+      end
+
+      # Apply the touches that were delayed. We're in a transaction already so there's no need to open one.
+      def apply
```
This could be (note the rename):
```ruby
# Applies delayed touches within the current transaction.
def apply_touches
  begin
    state.get_and_clear_records.each do |(klass, attrs), records|
      touch_records klass, attrs, records
    end
  end while state.more_records?
ensure
  state.clear_already_updated_records
end
```

Why do we have to use the `begin ... while` form instead of `while ...`?

BMorearty replied:
Beautiful refactor, @kaspth. I've made this change and moved the `while` to the top.

kaspth commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+
+      protected
+
+      attr_accessor :records, :already_updated_records
```
Can they simply be `attr_reader`?

BMorearty replied:
Yes. Done.

kaspth commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+        @records[[record.class, columns]] += [record] unless @already_updated_records[[record.class, columns]].include?(record)
+      end
+
+      def clear_records
```
This isn't used anywhere.

BMorearty replied:
Whoops. Good catch.Removed.

kaspth commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+      # As a side-effect, clears out the list of records.
+      def get_and_clear_records
+        records = @records
+        @records = Hash.new { SortedSet.new }
```
Either inline `clear_records` or add a call to it here :)

kaspth replied:
Ah, wait that won't work. Never mind.

kaspth commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+        @records.present?
+      end
+
+      def add_record(record, columns)
```
How about this :smile:
```ruby
def add_record(record, extra_columns)
  key = [
    record.class,
    record.timestamp_attributes_for_update_in_model.concat(extra_columns).map(&:to_sym).sort
  ]

  @records[key] += [record] unless @already_updated_records[key].include?(record)
end
```

`timestamp_attributes_for_update_in_model` uses `select` which returns an empty array if there's nothing to select. This means we can count on the return value responding to `concat` and just use that regardless.

Also it bothers me a bit the value for the key is wrapped in an array `[record]`. Doesn't `SortedSet`s have a method to add a single item without needing it to be an array or enumerable?

BMorearty replied:
I like some parts of the suggested refactor, but I find the `record.timestamp_attributes_for_update_in_model.concat(extra_columns).map(&:to_sym).sort` line to be a bit much to take in all at once.

I'd like to change it to this:

```ruby
      def add_record(record, columns)
        # Include the standard updated_at column and any additional specified columns
        columns += record.timestamp_attributes_for_update_in_model
        columns = columns.map(&:to_sym).sort

        key = [record.class, columns]
        @records[key] += [record] unless @already_updated_records[key].include?(record)
      end
```
> Also it bothers me a bit the value for the key is wrapped in an array [record]. 

Originally I did this:

```ruby
@records[key] << record unless ...
```

However, it doesn't work because of [the way Ruby handles default Hash values](http://stackoverflow.com/questions/16159370/ruby-hash-default-value-behavior). I think it's fine just to leave it as is.

kaspth commented on activerecord/lib/active_record/delay_touching/state.rb:
```diff
@@ -0,0 +1,71 @@
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
```
Come to think of it this whole business of keeping track of where we are in the update phase for a set of records feels weird to me.

`State` is of two minds; it manages past (`@already_updated_records`) and future (`@records`).

Perhaps it would be better to have `State` act as a linked list.
You could get rid of `@already_updated_records` (+ all that mutation business) and only refer to `@records`. When this state's records have been updated, you simple move on to the next state in the list :grin:

What do you think?

I remember @tenderlove doing something rather elegant like this in Action Dispatch, but I can't find it.

kaspth replied:
Here's the commit I was thinking of: https://github.com/rails/rails/commit/dc3f25c8a5aa64de9225f11498a389a2d31e880a

The code Aaron worked on has similar characteristics to this one. He let the `Scope` class handle a single level and then move back and forth using `parent` and `new` (an instance method) respectively.

BMorearty replied:
@kaspth I don't think a linked list is necessarily better in this case. There are a couple of reasons:

1. The code might exit and re-enter nested transactions. If I understand your suggestion correctly, a linked list would end up having to have branches:

    ```ruby
    transaction do
      transaction do
      end
      transaction do
      end
    end
    ```
2. It's true that `@already_updated_records` is only necessary when there's just one `State` left on the stack, because that's when we start touching records and keeping track of those touches. However, if I move it out to another instance of `State` (which I think is what you're suggesting), it's harder to keep track of. The `touch` override has to check it, which requires storing it as another thread-local variable. I think it will get messy.

As I wrote this answer, though, I did begin to realize that at the very least, I should rename `@already_updated_records` to `@already_touched_records`. But I realize that doesn't address your concerns about keeping track of two states at once.

kaspth replied:
I've had some more time to think about this and you're right the linked list wouldn't work. Thanks for your detailed explanation :heart:

BMorearty committed:
Changes based on code review.

BMorearty committed:
Merge remote-tracking branch 'upstream/master' into delay_touching

Conflicts:
	activerecord/CHANGELOG.md

BMorearty committed:
Rename already_updated_records => already_touched_records.

BMorearty commented:
Thanks for all the great feedback. Especially you, @kaspth.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+      delegate :add_record, to: :current_state
+
+      def delay_touching?
+        states.present?
```
Come to think of it `any?` might be better here :smile:

BMorearty replied:
I'd rather leave this as `present?` if you don't mind. It's a standard part of ActiveSupport, I think it's just as readable, and it doesn't require calling a block.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+      # touched records with the outer transaction's touched records.
+      def merge_transactions
+        num_states = states.length
+        states[num_states - 2].merge!(current_state) if num_states > 1
```
What happened to `states[-2].merge!(current_state) if delay_touching?` :smile:

BMorearty replied:
Oops, I missed the `[-2]`. I'll fix that. 

I can't use `if delay_touching?` because that will return true even when there is only one state on the stack. I need to make sure there are two or more states.

kaspth replied:
Right :+1:

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+        end
+      ensure
+        current_state.clear_already_touched_records
+      end
```
`apply_touches` will only be called when there's a single `State` on the stack. It also seems to know a lot about the `current_state`, so perhaps we should push this behavior into the state.

Then `start` could be `current_state.apply_touches if states.length == 1` instead. What do you think? :smile:

kaspth replied:
Whoops, I forgot to ask if using a `while` loop here is needed (as opposed to an `if` statement), since `apply_touches` is called with one state left? As far as I can tell there won't suddenly be more records after `touch_records` is called.

BMorearty replied:
> ... so perhaps we should push this behavior into the state.

My intent was that the state doesn't do the actual touching, it just keeps track of which records need to be touched and have already been touched. It's there as a helper for the main logic, which is in the DelayTouching module. If State does the actual touching, the name "State" seems wrong. I think it's better as is.

> I forgot to ask if using a `while` loop here is needed (as opposed to an `if` statement)....

An `if` statement won't work here. (I verified this by trying it.) The reason for the `while` loop is that each time an `after_touch` callback is called, it can have a side-effect of causing more calls to `touch` on other objects. (E.g., `belongs_to :something, touch: true`.) The `while` loop in `apply_touches` is there so that we can keep on iterating until there are no more touches caused by side-effects.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+            record.instance_eval do
+              attrs.each { |column| write_attribute column, current_time }
+              increment_lock if locking_enabled?
+              @changed_attributes.except!(*attrs)
```
@sgrif is there a better way to accomplish this without reaching into an ivar?

sgrif replied:
Yup. http://api.rubyonrails.org/classes/ActiveModel/Dirty.html#method-i-clear_attribute_changes

BMorearty replied:
Thanks. When I first wrote this code last summer for Rails 4, I don't think `clear_attribute_changes` existed. Or if it did, `touch` wasn't calling it. I'll update it.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+      end
+
+      # Touch the specified records--non-empty set of instances of the same class.
+      def touch_records(klass, attrs, records)
```
There's an inconsistency here with `attrs`. Some places you say `columns` versus `attrs` in others. This confused me because I kept thinking this was an `attrs` hash with values instead of an array of time columns to be updated.

BMorearty replied:
Oh yeah, good point. I'll change it to columns here and in `State`.

kaspth commented on activerecord/lib/active_record/delay_touching.rb:
```diff
@@ -0,0 +1,115 @@
+        states.push State.new
+        result = yield
+        apply_touches if states.length == 1
+        result
```
I would probably tap here:

```ruby
yield.tap do
  apply_touches if states.length == 1
end
```

BMorearty replied:
I avoided tap on purpose. I felt it was less readable and takes the same number of lines (if using do/end, not {}).

sgrif replied:
It's less about the number of lines, and more about clarifying intent. I can immediately see that the return value will be `yield`, and you just need to do something else before you return.

BMorearty replied:
@sgrif If you really want `tap`, I'll change it. But we have a difference of opinion here. I believe `tap` obfuscates the intent.

BMorearty replied:
Awaiting a further reply before proceeding--will you allow the way I think is clearer (no tap), or shall I use the way you think is clearer (tap)? :smiley:

kaspth replied:
As long as you tap into your :heart: when programming, either way is fine :smile:

BMorearty replied:
:heart:

stuarthannig replied:
II feel tap is more appropriate because it consolidates the logic in a single idea, compared to having to worry about the result variable showing up twice. Same amount of lines, but to me I understand what I am getting out of the code with the do block.

BMorearty committed:
Changes based on code review.

BMorearty committed:
Consistently use the term 'columns', not 'attrs'.

BMorearty commented:
I've applied all your recommendations except the ones I pushed back on.

jeremy commented:
@BMorearty Consider that we may not need touch batching, but just touch coalescing, to alleviate the primary pains with `#touch` — and that could significantly reduce our implementation complexity.

The big pains are deadlocks and lock contention on parent records. Since `#touch` ordering is unpredictable and scattered throughout the transaction, we suffer deadlocks that are difficult to diagnose and eliminate. Since `#touch` is repeated on parent records, we suffer long exclusive locks on high-contention records, blocking concurrent transactions on the same parents.

We can nail both of these by marking records when they're touched, then performing the actual touch (the same one AR would immediately issue) before commit. Here's a sketch from a plugin form of this: https://gist.github.com/jeremy/84134ad08f2a137aa1ef

This ensures that the record gets an updated timestamp immediately, that repeated touches are coalesced into a single touch per record, and that touches are performed in a predictable order at the end of the transaction. It depends on couple of AR improvements: before_commit #18903, commit callbacks with manual record enrollment #18904, and touching with an explicit `:time` #18905.

These improvements equip us with a toolbox to make a straightforward implementation of touch coalescing and go on to reuse the same tools to do other kinds of activity coalescing within AR, plugins, and apps.

What do you think about starting with touch coalescing only, then expanding to batching as a second exploration that builds on that?

BMorearty commented:
@jeremy Thanks for the feedback. Here are my thoughts.

> The big pains are deadlocks and lock contention on parent records.

These were not the big pains we experienced on a Rails project last summer that prompted me to write the [activerecord-delay_touching gem](https://github.com/godaddy/activerecord-delay_touching/), on which this PR is based. 

We didn't run across lock contention. For us, the big pain was the large amount of time it would take to touch a lot of records. There was a huge cascade of touches. Some of them would be fixed with your proposal. For example, when updating a lot of Variants (this was a Spree-based project), the associated Product was touched many times. Your proposal would reduce that to one touch for the Product. However, Spree has a complex model that also causes it to touch multiple objects in the same table. One example: a Product can belong to multiple Categories. When the Product is touched, all associated Categories are touched. 

Anyway, that's the background on why I wrote this the way I did, not only coalescing touches on a single record but batching the touches that occur on multiple records in a table.

dhh commented:
We could coalesce many updates to one model type at the end as well. I don't think that precludes that. When we mark "this needs updating", we could mark that as `touches[Product] << id` or whatever, and then the final execution of the touching is clever.

dhh commented:
I think the main simplification is to move to a mark-and-bulk-execution-at-the-end model.

BMorearty commented:
@jeremy Your gist is certainly simpler than the code in this PR. 

Part of the reason for the complexity in the PR was to handle nested transactions as they're documented--by which I mean, a rollback from an inner `:requires_new` transaction will not update the timestamp on the records in the inner transaction, and a rollback from an inner non-`:requires_new` transaction will still touch them if the outer is committed. I'm not sure if this requirement will be met with the code in the gist.

Another, more important requirement that cause a bit of complexity was that side-effects (more touches) from `after_touch` will also be coalesced. It doesn't look to me like the gist addresses that. The first level of touches and `after_touch` hooks will execute, but if they cause more touches, I don't think those will execute.

If we eliminate requirements like those, and eliminate the ability to coalesce all touches in a single table, then yes, the code can be simpler, as the gist shows. I'm certain we don't want to eliminate the side-effects requirement, though, right? Because then `belongs_to` with `touch: true` will only bubble up one level.

(Or we could still coalesce touches in a single table, as @dhh suggested, but that would re-introduce some of the complexity in this PR. E.g., go back to keying off not only the class, but also the columns that have to be touched. And either live with the timestamps having different values in memory than they have on the DB, or update memory and the DB together, as I did.)

jeremy commented:
@BMorearty `after_touch` will bubble once when `#touch_now` is called before commit. That also resolves some long-standing surprises with this callback, e.g. #8759.

Re. performance, the Spree benchmark definitely show the best of batching: small number of distinct models (2), large number of distinct parent/grandparent records to touch (150). Coalescing offers no savings when we aren't re-touching common parent/grandparent records. In Russian-doll hierarchies, the # of distinct touchable records is much smaller. Updating N child records touches the same parent and grandparent record N times. Coalescing reduces the Russian-doll case to 2 updates, same as batching. Possible examples in Spree are updating order line item quantities (1 order touch) or backing up a viewable's assets (1 viewable touch).

Re. transaction sync, `before_commit` fires when committing a non-joinable transaction — the outermost txn or an inner `requires_new` txn — so rollback/commit "just works" with the existing transaction state sync machinery.

BMorearty commented:
> after_touch will bubble once when #touch_now is called before commit.

Once isn't enough, right? We'd need to bubble it all the way up.

I think my use of the word "batch" in the title of this PR may have caused some confusion about its purpose. I just want to clarify that the primary purpose of this code--the reason I wrote it--is the same as the primary purpose you described: the `touch: true` case that causes one record to be touched over and over. The fact that my code also touches multiple records in a single round-trip is secondary, for the Spree case.

jeremy commented:
Right, I mean "once" in the "when" sense, ha!

We're on the same page re. purpose :grin:

On Wednesday, February 11, 2015, Brian Morearty <notifications@github.com>
wrote:

> after_touch will bubble once when #touch_now is called before commit.
>
> Once isn't enough, right? We'd need to bubble it all the way up.
>
> I think my use of the word "batch" in the title of this PR may have caused
> some confusion about its purpose. I just want to clarify that the primary
> purpose of this code--the reason I wrote it--is the same as the primary
> purpose you described: the touch: true case that causes one record to be
> touched over and over. The fact that my code also touches multiple records
> in a single round-trip is secondary, for the Spree case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/18824#issuecomment-74017115>.
>

dhh commented:
What's the next step for this? @BMorearty is it clear how to move forward with the before_commit approach?

BMorearty commented:
Yes, @dhh, I think it's clear, but it sounds like this approach is completely different from the approach in the gem I wrote. I originally volunteered because I had already written the code and it had been working in production for a while (and it was [an improvement](https://signalvnoise.com/posts/3850-are-we-making-it-better) over the current behavior of touches). If someone else wants to run with this new approach, I'm totally fine with that. 

By the way, I'm still not convinced the before_commit approach--as written in the gist--will correctly bubble up all side-effect touches to the Nth degree. It would need to be tested for that.

dhh commented:
I understand, Brian. Thanks for your work on this so far. Let’s open the issue up again to other contributors to take a swing at the before_commit approach. If that doesn’t pan out, I’m sure we’ll be back on this train ;)


> On Feb 20, 2015, at 09:59, Brian Morearty <notifications@github.com> wrote:
> 
> Yes, @dhh, I think it's clear, but it sounds like this approach is completely different from the approach in the gem I wrote. I originally volunteered because I had already written the code and it had been working in production for a while (and it was an improvement over the current behavior of touches). If someone else wants to run with this new approach, I'm totally fine with that.
> 
> By the way, I'm still not convinced the before_commit approach--as written in the gist--will correctly bubble up all side-effect touches to the Nth degree. It would need to be tested for that.
> 
> —
> Reply to this email directly or view it on GitHub.
>

BMorearty commented:
:+1:

dhh commented:
@arthurnn is going to give it a go on the before_commit approach. We'll reconvene in a new PR when it's ready.

kaspth commented:
@BMorearty thanks, Brian. I hope you enjoyed the reviews :smile:
